vllm/compilation/pass_manager.py

```diff
@@ -0,0 +1,82 @@
+from typing import Sequence
+
+from torch import fx as fx
+from torch._inductor.codecache import BypassFxGraphCache
+
+from vllm.logger import init_logger
+
+from .config import PassConfig
+from .functionalization import FixFunctionalizationPass
+from .fusion import FusionPass
+from .inductor_pass import InductorPass, InductorPassType
+from .reshapes import RedundantReshapesPass
+
+logger = init_logger(__name__)
+
+
+class PostGradPassManager:
+    """
+    The pass manager for post-grad passes.
+    It handles configuration, adding custom passes, and running passes.
+    It also supports pickling, which is used by the Inductor code cache.
+    TODO in torch==2.6, use CustomGraphPass
+    (torch._inductor.custom_graph_pass.CustomGraphPass)
+
+    The order of the post-grad post-passes is:
+    1. passes (constructor parameter)
+    2. default passes (RedundantReshapesPass, FusionPass)
+    3. config["post_grad_custom_post_pass"] (if it exists)
+    4. fix_functionalization
+    This way, all passes operate on a functionalized graph.
+    """
+
+    def __init__(self, passes: Sequence[InductorPassType] = ()):
+        self.passes = list(passes)
+
+    def __call__(self, graph: fx.Graph):
+        for pass_ in self.passes:
+            pass_(graph)
+
+        # always run fix_functionalization last
+        self.fix_functionalization(graph)
+
+    def configure(self, pass_config: PassConfig):
+        self.pass_config = pass_config
+        if pass_config.enable_reshape:
+            self.passes += [RedundantReshapesPass(pass_config)]
+
+        if pass_config.enable_fusion:
+            self.passes += [FusionPass.instance(pass_config)]
+
+        self.fix_functionalization = FixFunctionalizationPass(pass_config)
+
+    def add(self, pass_: InductorPassType):
```

Comment:
I think by using qualified name for the passes, we can avoid the serialization problem.

Reply:
One downside of using qualified name, is it does not support partially bound functions, e.g. if a pass is something like `partial(general_transform_func, args)` . I don't know if we have this kind of use cases right now.

Reply:
The issue with that is that if a user changes the implementation for a function, the qualified name stays the same and the inductor cache will use the outdated code, even though it should recompile with the new function implementation.

Reply:
I think that is fine. that's only for developers. and we can issue a warning to tell developers that they should always recompile when they are developing the code.

Reply:
I think that would add verbosity to the warnings and make the workflow more complicated: when working on passes, developers would have to manually clear the cache. This approach doesn't hurt any particular situation.

Unless you're talking about the case where outside developers supply their own passes - but there they can also just not inherit from InductorPass and the cache will always be bypassed. Or they can inherit and provide their own UUID which doesn't change and deal with caching themselves. Or they set it properly and get proper caching.

Reply:
yeah I'm talking about the case where outside developers supply their own passes. It makes sense for me to clear the cache if you change the compilation setting. and we will add per-config compilation cache later, and only use cache when users specify it (by default not using cache from previous run)